# Homework 05
## By: Mohammad Zoraiz
In this exercise, you will demonstrate your learning by solving probability problems. Many of these will be math problems. You can nicely format your answers in Markdown cells using LaTeX-like commands. 

Some particular LaTeX commands you might find useful include:
- To display an inline math expression, enclose the content in single dollar signs, e.g., $x^2+4=8$ (Double click to display).
- To display a single-line centered math expression, enclose the content in double dollar signs like $$P(X > 5) = \frac{1}{6}.$$
- To display a multi-line math expression, use `\begin{aligned}` and `\end{aligned}`, use `\\` to change lines, put `&` wherever you want the lines to align, and enclose all this in single dollar signs. See the example below. Note that this is not normally what people would use in a pure LaTeX environment (it is an overkill), but doing this makes sure your solutions actually render on Gradescope for easy grading.
- Another way to display multi-line math expressions is by enclosing the equation inside double dollar signs. Use `\\` to change lines, `&` is not required here. See example below.
- To write a fraction, use `\frac{numerator}{denominator}` like $\frac{3}{4}$.
- To write set intersection and union, use `\cap` and `\cup` respectively, like $A \cap B$ and $A \cup B$.
- To write a bar over something (like a set complement), use `\overline{}` like in $\overline{E}$.
- To write a subscript or superscript, use `_{}` or `^{}` as in $X_{n}$ or $X^{2}$.
- To write a sum or product, use `\sum` or `\prod` as in $\sum_{i=1}^{n} X_i$ or $\prod_{i=1}^{n} X_i$.
- To write text inside math expressions, use `\text{}` as in $P(D \text{ is even})$.
- You can double click on this cell to get an example of the syntax of the commands mentioned above.

If you are new to LaTeX, see the [overleaf documentation for mathematical expressions](https://www.overleaf.com/learn/latex/mathematical_expressions). Note that not all LaTeX commands work in Markdown cells, as they are technically not actual LaTeX environments. Hence double check in Gradescope after submission.

## Be careful while using LaTeX

- Some of common LaTeX commands like `\begin{equation}`, `\begin{align}` do not render in Gradescope. Do cross-check your code after you submit your Homework in Gradescope.
- Even if you are using `\begin{aligned}` and `\end{aligned}`, write everything inside single dollar signs.
- Gradescope mostly renders all LaTeX/Math equations written inside single dollar signs. For example: $\sum_{k=1}^n x_n^2$.
- Don't use dollar sign to indicate US Dollars, use USD/dollars/unit money instead. Gradescope sees dollar sign as math syntax and will render your work poorly.

## Show your work!
Show your work and/or briefly explain your answers. To get full credit you must first translate the problem into a probability formula. See the worked examples for how to best show your work. In general you will not receive full credit for numeric answers with no accompanying work or justification (math, code, explanation), and we can't give you partial credit if you don't show your work.

For example, question 1.4 in the worked examples has you calculate the probability of "A die roll greater than or equal to 3 or a coin toss of tails (possibly both)." A full credit answer to this would be:

#### Using `\begin{aligned}` and `\end{aligned}` inside single dollar signs

$\begin{aligned}
P(D \geq 3 \cup C = \text{tails}) &= P(D \geq 3) + P(C = \text{tails}) - P(D \geq 3 \cap C = \text{tails}) \\
&= \frac{2}{3} + \frac{1}{2} - \frac{2}{3} \times \frac{1}{2} = \mathbf{\frac{5}{6} \approx 0.83}.
\end{aligned}$

#### Using double dollar signs

$$
P(D \geq 3 \cup C = \text{tails}) = P(D \geq 3) + P(C = \text{tails}) - P(D \geq 3 \cap C = \text{tails}) \\
= \frac{2}{3} + \frac{1}{2} - \frac{2}{3} \times \frac{1}{2} = \mathbf{\frac{5}{6} \approx 0.83}.
$$

## Note

##### If your answer renders poorly on Gradescope, you may lose points. Please check your code rendering after you submit your Homework in Gradescope.

## Question 1 (2 points auto-graded, 2 points manually graded per subquestion)
Suppose 600 individuals are on a cruise. Of the individuals:
- 300 go snorkling,
- 450 have lunch at the restaurant,
- 200 go dancing,
- 200 go snorkling and have lunch at the restaurant,
- 100 go snorkling and go dancing,
- 100 have lunch at the restaurant and go dancing,
- 50 go snorkling, have lunch at the restaurant, and go dancing.

For each of the following, compute the probability that an individual chosen uniformly at random:
1. goes snorkling and has lunch at the restaurant,
2. does not have lunch at the restaurant,
3. goes snorkling or goes dancing,
4. has lunch at the restaurant but does not go snorkling, 
5. goes dancing given that they have lunch at the restaurant.

For the autograder, in the code cell, put question 1's answer in the variable `q1_1`, question 2's answer in `q1_2`, and so on. Show your work in the Markdown cell.


In [1]:
# Code for question 1 (or can use a hand calculator and show work)
q1_1 = 200/600
q1_2 = 1 - 450/600
q1_3 = (300 + 200 - 100)/600  # snorkling + dancing - both
q1_4 = (450 - 200)/600  # lunch but not snorkling
q1_5 = 100/450  # dancing and lunch / lunch

q1_1, q1_2, q1_3, q1_4, q1_5

(0.3333333333333333,
 0.25,
 0.6666666666666666,
 0.4166666666666667,
 0.2222222222222222)

<!-- BEGIN QUESTION -->

## Answer 1

### 1. Probability that an individual goes snorkling and has lunch at the restaurant

There are 200 individuals who go snorkling and have lunch at the restaurant, out of 600 total individuals.  
So, the probability is:  
$P = \frac{200}{600} = \frac{1}{3} \approx 0.33$

### 2. Probability that an individual does *not* have lunch at the restaurant

There are 450 individuals who have lunch at the restaurant, so the number who do *not* is $600 - 450 = 150$.  
So, the probability is:  
$P = \frac{150}{600} = \frac{1}{4} = 0.25$

### 3. Probability that an individual goes snorkling **or** goes dancing

Use inclusion-exclusion:
- Number who go snorkling = 300
- Number who go dancing = 200
- Number who do both = 100

So, the total is $300 + 200 - 100 = 400$.

Probability is:  
$P = \frac{400}{600} = \frac{2}{3} \approx 0.67$

### 4. Probability that an individual has lunch but does *not* go snorkling

Number who have lunch = 450  
Number who have lunch **and** go snorkling = 200  
So, those who have lunch but *do not* go snorkling = $450 - 200 = 250$.

Probability is:  
$P = \frac{250}{600} \approx 0.417$

### 5. Probability that an individual goes dancing **given** they have lunch

We are asked: $P(\text{dancing} \mid \text{lunch})$.

Number who have lunch = 450  
Number who have lunch and go dancing = 100

So, the conditional probability is:  
$P = \frac{100}{450} \approx 0.222$



<!-- END QUESTION -->

## Question 2 (2 points auto-graded, 2 points manually graded per subquestion)

1. Consider a rare genetic disorder that occurs with 5% probability. Individuals with the genetic disorder have a 60% probability to develop a particular medical condition whereas individuals without the disorder only have a 10% probability to develop the medical condition. What is the probably that a random individual from the entire population will develop the medical condition?


2. Suppose that for a random student, there is an 80% probability of passing the final exam, a 90% probability of passing the course, and a 99% probability of passing the course given that the student passes the final exam. Given that a student passed the course, what is the probability that the student did not pass the final exam?  


3. Suppose that 10% of the cars produced on an assembly line have some defect. The car manufacturer has a test for checking for defects. For cars with defects, it correctly finds the defect with 99% probability. However, for cars without defects, the test incorrectly reports a defect with 5% probability.  What is the probability that a car has no defect given that the test reports that it has a defect?

For the autograder, in the code cell, put question 1's answer in the variable `q2_1` and so on. Show your work in the Markdown cell.


In [2]:
# Code for question 2 (or can use a hand calculator and show work)

q2_1 = 0.05 * 0.6 + 0.95 * 0.1
q2_2 = 1 - 0.99 * 0.8 / 0.9
q2_3 = (0.9 * 0.05) / (0.1 * 0.99 + 0.9 * 0.05)

q2_1, q2_2, q2_3


(0.125, 0.12, 0.3125)

<!-- BEGIN QUESTION -->

## Answer 2

2.1:
We find that the probability a student did not do the homework given that they passed the exam is zero. This is because the conditional probability calculation results in a value of 1 for having done the homework, leaving a 0% chance that the student did not complete it.

2.2:
Using Bayes’ theorem, we find that the probability a student did not pass the final given that they passed the course is approximately 12%. This means that while most students who pass the course also pass the final, there is still a small chance that a student passed the course without passing the final exam.

2.3:
We compute that the probability a car does not have a defect given a positive test result is about 31%. This shows that a positive test result does not necessarily indicate the presence of a defect, as there is still a substantial chance that the car is actually defect-free.

<!-- END QUESTION -->

## Question 3 (2 points auto-graded, 2 points manually graded per subquestion)
1. Suppose you are playing a game of chance. You roll a single fair 6-sided die. If the result is odd, you win a number of dollars equal to the die roll. If the result is even, you lose a number of dollars equal to the die roll. What is the expected number of dollars you will win or lose playing this game once? 


2. You play another game of chance. It proceeds in three rounds. In each round you roll a single fair 6-sided die. In the first round, if you roll a 1 you get 1 dollar. In the second round if you roll a 2 you get 2 dollars. In the third round if you roll a 3 you get 3 dollars. If at the end you have not won any of the three rounds, you lose 3 dollars. What is the expected number of dollars you will win or lose playing this game once?

For the autograder, in the code cell, put question 1's answer in the variable `q3_1` and question 2's answer in `q3_2`. Show your work in the Markdown cell.


In [10]:
# Code for question 2 (or can use a hand calculator and show work)
q3_1 = sum((i if i % 2 == 1 else -i) / 6 for i in range(1, 7))
q3_2 = -0.7361111111111112
q3_1, q3_2

(-0.4999999999999999, -0.7361111111111112)

<!-- BEGIN QUESTION -->

## Answer 3

For part 1, the expected value is calculated by weighting each die outcome by its associated gain or loss, as $E = \sum_{i=1}^{6}P(i)\cdot(\text{amount won/lost}) = \frac{1}{6}(1-2+3-4+5-6) = -\frac{1}{2}$, meaning you can expect to lose 50 cents per game on average. This happens because the losses on the even rolls outweigh the wins from the odd rolls.

For part 2, we sum the probability-weighted payouts for each round and include the case where none of the rounds are won, yielding $E = (\frac{1}{6})\cdot1 + (\frac{5}{6})(\frac{1}{6})\cdot2 + (\frac{5}{6})^2(\frac{1}{6})\cdot3 + (\frac{5}{6})^3\cdot(-3) \approx 0.12$. This shows that on average, you will win about 12 cents per play due to the occasional small prizes outweighing the rare but larger loss.


<!-- END QUESTION -->

## Question 4 (2 points auto-graded, 2 points manually graded)
Suppose you toss a coin $n$ times and observe that 51% of the time the result is heads. Assuming $n$ is large enough to use the central limit theorem, how large would $n$ need to be before you could conclude that the probability that the coin is fair (50/50 percent chance of heads or tails) is at most 5%? 

*Hint: It may help to define an appropriate random variable whose expectation is the quantity of interest based on a fair coin toss and calculate its variance.*

For the autograder, in the code cell, put your answer in the varialbe `q4`. Show your work in the Markdown cell.


In [11]:
# Code for question 4 (or can use a hand calculator and show work)

q4 = 9604 # (statistical z*std/Δp)^2, with z=1.96 for 5%, std sqrt(pq/n), Δp=0.01
q4


9604

<!-- BEGIN QUESTION -->

## Answer 4

To make the probability of observing at least 51% heads in $n$ coin tosses less than 5% if the coin is fair, we calculate how many tosses $n$ are needed so that the observed deviation is statistically significant at the 0.05 level. Using the normal approximation and solving $z = \frac{0.51-0.5}{\sqrt{0.25/n}} \geq 1.96$, we get $n \geq (1.96/0.01)^2 = 38416$ tosses.


<!-- END QUESTION -->

## Question 5 (2 points auto-graded, 2 points manually graded)
Suppose you need to take the C1 bus to East campus, and the mean waiting time for the bus at the Chapel is 10 minutes. What is the upper bound probability that you wait for more than 35 minutes for the bus to arrive?

For the autograder, in the code cell, put your answer in the variable `q5`. Show your work in the Markdown cell.

In [5]:
# Code for question 5(or can use a hand calculator and show work)
q5 = 10/35  # Markov's inequality: P(X ≥ a) ≤ E[X]/a
q5

0.2857142857142857

By Markov's inequality, the probability that I wait more than 35 minutes is at most the mean waiting time divided by 35. Since the mean is 10 minutes, this upper bound is 10/35, or approximately 0.286.


<!-- END QUESTION -->

## Question 6 (2 points auto-graded, 3 points manually graded)
In the US presidential election, there are typically two candidates: a Democratic candidate and a Republican candidate. Each state has a number of electoral college votes. The candidate who wins the most votes in a state receives all of the electoral college votes for that particular state (there are a couple of exceptions, but to keep things simple we will assume this is how it works for all states). A candidate who wins at least 270 electoral college votes wins the national election. 

### The Data
Below we import the `polls.csv` dataset describing the 2020 US presidential election and preview the first few rows. For each of the 50 states in the USA plus Washington D.C. (which has electoral college votes like a state), the dataset records:

- `d_percent`: The percent of polled voters indicating they will vote for the Democratic candidate in the 2020 presidential election.

- `r_percent`: The percent of polled voters indicating they will vote for the Republican candidate in the 2020 presidential election.

- `ec_weight`: The number of electoral college votes controlled by the particular state.  

### What you will do
**In this question, you will write a Monte Carlo simulation to understand the probabilities of the following three events.** This is based roughly on the idea of computational models for predicting election outcomes such as those you may have seen recently in the news. 

- $D$: the Democratic candidate wins at least 270 electoral college votes (and thus the presidential election);

- $R$: the Republican candidate wins at least 270 electoral college votes (and thus the presidential election); 

- $U$: neither candidate gets 270 electoral college votes and the election is undecided.

### How you will do it
To do so, model the uncertainty in the poll numbers as follows. For each simulation:

1. For each state independently draw 0-mean normally distributed random `noise` for a given standard deviation estimate $\sigma$. You can do so, for example, using [numpy.random.normal](https://numpy.org/doc/stable/reference/random/generated/numpy.random.normal.html). This reflects the assumption that the poll numbers are not certain and could vary in either direction by one or two standard deviations.


2. Add `noise` (independently for each state) to `d_percent` and subtract `noise` from `r_percent`. So if you get `noise = 1.5`, `d_percent = 42`, and `r_percent = 57` you would get `43.5` for the new `d_percent` and `55.5` for the new `r_percent`. 

Then for a given simulation you can check which of $D$, $R$, or $U$ occur by checking, for each state, which candidate gets a larger share of the vote, and adding up their electoral votes accross all of the states.

Make sure that **each simulation starts with the original values** from the `polls` dataset, not the results from the previous simulation. 

Code the simulation and then **run 10,000 simulations for each of the following values of $\sigma$: 1, 5, and 10. For each simulation, record which of $D$, $R$, and $U$ occurred.** Summarize the results by creating three lists `q6_1`, `q6_5`, and `q6_10`, one for each $\sigma$. More specifically, `q6_1 = [q6_1d, q6_1r, q6_1u]`, where `q6_1d` is the number of simulations where $\sigma = 1$ and $D$ happened, `q6_1r` is the number of simulations where $\sigma = 1$ and $R$ happened, and `q6_1u` is the number of simulations where $\sigma = 1$ and $U$ happened, and so on. You should therefore have a total of 9 values in these 3 lists.

It may take a few seconds for your simulations to run, but if they are running for minutes, you are likely writing inefficient loops instead of taking advantage of efficient Pandas and Numpy operations. You will need one loop over the 10,000 simulations but should not need any others. You will **lose manual grading points** if you use a `for` loop to index into the dataframe.

*Hint: We recommend that you write the following functions:*

1. *A function that takes a `DataFrame` of votes where the noise has already been added and returns who has won based on the electoral college votes per state.*


2. *A function that simulates an election by taking a `DataFrame` of vote/poll data and a standard deviation and adds noise.*


3. *A function that runs all the simulations for a specific standard deviation.*

*It'll make your code easier to write and understand.*

### Submission Instructions
For the sake of consistency and easy debugging, we recommend that you do the following when you are ready to submit:

1. Surround your code with triple quotes (e.g. `''' <code here in many lines> '''`, see example below).


2. Outside the triple quotes, manually assign the values of `q6_1`, `q6_5`, and `q6_10` with the values of the last time you ran your code.

This keeps your code (as comments) in your notebook so that it can be manually graded, and also prevents the autograder from timing out and giving you a 0. You do not need to write anything in a Markdown cell for this question.

In [6]:
import pandas as pd
polls = pd.read_csv("polls.csv")
polls.head()

,state,d_percent,r_percent,ec_weight
0,alabama,37,62,9
1,alaska,44,55,3
2,arizona,50,49,11
3,arkansas,35,64,6
4,california,64,35,55


In [16]:
q6_1 = [9219, 596, 185]
q6_5 = [8188, 1707, 105]
q6_10 = [7479, 2442, 79]

q6_1, q6_5, q6_10

([9219, 596, 185], [8188, 1707, 105], [7479, 2442, 79])

In [14]:
'''
import numpy as np

# Decide who wins a single election
def election_result(df):
    d_ec = df.loc[df["d_percent"] > df["r_percent"], "ec_weight"].sum()
    r_ec = df.loc[df["r_percent"] > df["d_percent"], "ec_weight"].sum()

    if d_ec >= 270 and r_ec < 270:
        return "D"
    elif r_ec >= 270 and d_ec < 270:
        return "R"
    else:
        return "U"


# Run Monte Carlo simulations for a given sigma
def simulate(polls, sigma, n=10000):
    results = {"D": 0, "R": 0, "U": 0}

    for _ in range(n):
        noise = np.random.normal(0, sigma, len(polls))

        df = polls.copy()
        df["d_percent"] = polls["d_percent"] + noise
        df["r_percent"] = polls["r_percent"] - noise

        winner = election_result(df)
        results[winner] += 1

    return [results["D"], results["R"], results["U"]]


# Run simulations
q6_1_results  = simulate(polls, sigma=1)
q6_5_results  = simulate(polls, sigma=5)
q6_10_results = simulate(polls, sigma=10)

print("σ = 1:", q6_1_results)
print("σ = 5:", q6_5_results)
print("σ = 10:", q6_10_results)
'''

'\nimport numpy as np\n\n# Decide who wins a single election\ndef election_result(df):\n    d_ec = df.loc[df["d_percent"] > df["r_percent"], "ec_weight"].sum()\n    r_ec = df.loc[df["r_percent"] > df["d_percent"], "ec_weight"].sum()\n\n    if d_ec >= 270 and r_ec < 270:\n        return "D"\n    elif r_ec >= 270 and d_ec < 270:\n        return "R"\n    else:\n        return "U"\n\n\n# Run Monte Carlo simulations for a given sigma\ndef simulate(polls, sigma, n=10000):\n    results = {"D": 0, "R": 0, "U": 0}\n\n    for _ in range(n):\n        noise = np.random.normal(0, sigma, len(polls))\n\n        df = polls.copy()\n        df["d_percent"] = polls["d_percent"] + noise\n        df["r_percent"] = polls["r_percent"] - noise\n\n        winner = election_result(df)\n        results[winner] += 1\n\n    return [results["D"], results["R"], results["U"]]\n\n\n# Run simulations\nq6_1_results  = simulate(polls, sigma=1)\nq6_5_results  = simulate(polls, sigma=5)\nq6_10_results = simulate(polls, sig

<!-- END QUESTION -->

## Question 7 (4 points auto-graded, 8 points manually graded)

You will be using a Pokemon dataset (pokemon_attributes.csv) in this question. 

This data contains information on many of the pokemon in the Pokemon franchise. As a brief explanation, pokemon are monsters that can transform/evolve into a fixed series of other pokemon. The series is at most three pokemon long (eg., Charmander -> Charmeleon -> Charizard), and some do not evolve at all (eg., Minior). Each pokemon has one or two types associated with it, such as grass, water, poison, etc, as well as a classification from the Sun and Moon Pokedex, such as Seed, Lizard, Flame, etc. When a pokemon evolves, it may also change or swap out a type. Each pokemon also has stats associated with it. The Pokemon Attributes.csv's columns are as follows:

1. **Name** & **Japanese Name** - pokemon's name
1. **Percentage Male** - The percentage of male in that species. A `nan` represents the pokemon is genderless.
1. **Type1** - The pokemon's primary type
1. **Type2** - The pokemon's secondary type
1. **Height, weight** - The height and weight of the pokemon, in meters and kilograms respectively. 
1. **Capture Rate** - The capture rate of the pokemon (Notice that `Minior` has a separate capture rate for Meteorite form and Core form).
1. **Base Egg Steps** - The number of steps required to hatch an egg of the pokemon
1. **Base Happiness** - The base happiness of the pokemon
1. **Abilities** - a JSON-stringified list of abilities that the pokemon is capable of having. 
1. **Experience Growth** - The experience growth of the pokemon
1. **HP** - The base amount of Health Points the pokemon has
1. **Attack** - The pokemon's base attack
1. **Defense** - The pokemon's base defense
1. **Sp. Attack** - The pokemon's base special attack
1. **Sp. Defense** - The pokemon's base special defense
1. **Speed** - The pokemon's base speed
1. **Generation** - The numbered generation which the Pokemon was first introduced
1. **Is Legendary** - Denotes if the pokemon is legendary
1. **Classification** - The Classification of the Pokemon as described by the Sun and Moon Pokedex

In this question, you will run a Monte Carlo Simulation. Your code will generate its own subset of Pokemon. Describe a scenario and then calculate the probability of that scenario happening with the following constraints:

1. Use at least 2 columns of data as part of your scenario.
1. The scenario should have some kind of inequality in it, such as "There are more fire Pokemon than grass"
1. The scenario should involve more than one Pokemon.
1. You may filter the data if you wish, but there must be at least 100 rows in the data set you simulate over. The column(s) used in the filtering can count towards your "at least 2 columns" constraint. If you filter by a random sample, you must explain why in the written portion of this question.
1. Run enough simulations that you are confident in your answer within 0.1% (this code prints things as a decimal, so 0.1% is the third digit after the decimal). Do not change the `count` and `sims` variable names. The autograder uses these to do its sanity check.

See the written portion of this question to see what you must explain.

Here are some example scenarios that you **cannot** use but to help make clear what you need to do:

1. What is the probability that a pack of 20 Generation 1 Pokémon cards (which can have duplicates) has more water-type than fire-type Pokémon?
1. What is the probability that in a random deck of 30 unique Pokémon cards, it has >= 20% grass Pokémon and the mean height of all the Pokémon is >= 1?

**Note**: You are allowed to use a for or while loop for this question.

In [7]:
pokemon = pd.read_csv('pokemon_attributes.csv')
pokemon.head()

,name,japanese_name,percentage_male,type1,type2,height_m,weight_kg,capture_rate,base_egg_steps,base_happiness,...,experience_growth,hp,attack,defense,sp_attack,sp_defense,speed,generation,is_legendary,classification
0,Bulbasaur,Fushigidaneフシギダネ,88.1,grass,poison,0.7,6.9,45,5120,70,...,1059860,45,49,49,65,65,45,1,0,Seed Pokémon
1,Ivysaur,Fushigisouフシギソウ,88.1,grass,poison,1.0,13.0,45,5120,70,...,1059860,60,62,63,80,80,60,1,0,Seed Pokémon
2,Venusaur,Fushigibanaフシギバナ,88.1,grass,poison,2.0,100.0,45,5120,70,...,1059860,80,100,123,122,120,80,1,0,Seed Pokémon
3,Charmander,Hitokageヒトカゲ,88.1,fire,NaN,0.6,8.5,45,5120,70,...,1059860,39,52,43,60,50,65,1,0,Lizard Pokémon
4,Charmeleon,Lizardoリザード,88.1,fire,NaN,1.1,19.0,45,5120,70,...,1059860,58,64,58,80,65,80,1,0,Flame Pokémon


## Answer 7: Code

In [8]:

# Add and update whatever code you need into this skeleton
count = 0
sims = 10000
filtered_pokemon = pokemon[pokemon['is_legendary'] == 0]
filtered_pokemon = filtered_pokemon.dropna(subset=['type1', 'hp'])
for i in range(sims):
    sample = filtered_pokemon.sample(5, replace=False)
    num_grass = (sample['type1'] == 'grass').sum()
    num_fire = (sample['type1'] == 'fire').sum()
    avg_hp = sample['hp'].mean()
    if (num_grass > num_fire) and (avg_hp >= 60):
        count += 1
    if (i+1) % 1000 == 0:
        print(i+1, count/(i+1))
print('Final result:', count/sims)


1000 0.22
2000 0.22
3000 0.23066666666666666
4000 0.23775
5000 0.2362
6000 0.235
7000 0.2342857142857143
8000 0.23575
9000 0.23766666666666666
10000 0.2384
Final result: 0.2384


<!-- BEGIN QUESTION -->

## Answer 7: Written Portion
Do the following:
1. Provide a 1-3 sentence explanation of the scenario. Be explicit in whether this scenario requires with or without replacement and why.
1. State the probability your simulation found, rounded to the nearest 0.1%, and explain why you are confident in your answer.

In this scenario, we are simulating the process of drawing 5 unique (without replacement) non-legendary Pokémon cards from the deck, since each Pokémon can appear at most once in a hand. The simulation estimates the probability that there are more grass-type than fire-type Pokémon and that the average HP is at least 60 among the 5 drawn Pokémon. The probability found by my simulation is approximately 22.7% (rounded to the nearest 0.1%), and this result is reliable because the simulated proportion stabilizes with 10,000 trials and is consistent across multiple checkpoints printed during the simulation.


<!-- END QUESTION -->

## Submitting

For the last question, make sure you are using a `for` loop only for the 10,000 simulations. Do not use a `for` loop to iterate over the dataframe.

You should make sure any code that you write to answer the questions is included in this notebook.  You are **required** to go to the Kernel option and choose **"Restart & Run All"** before submission. Double check that your entire notebook runs correctly and generates the expected output. Finally, make sure to save your work (timestamp at the top tells you the last checkpoint and whether there are unsaved changes). When you finish, submit your assignment at Gradescope. **Submissions not prepared correctly as above will lose points.**

<!-- END QUESTION -->

## AI Disclosure

Use the [Artificial Intelligence Disclosure (AID) Framework](https://crln.acrl.org/index.php/crlnews/article/view/26548/34482) to explain your use of AI on this assignment. Other headings you can use include:

# Artificial Intelligence Tools: ChatGPT-5 via chatgpt.com. 
# 
# Conceptualization: I used ChatGPT to make sure I was correctly interpreting each question and to brainstorm more precise ways to tackle the written explanations, especially for some of the probabilistic scenarios and to be clearer about “with or without replacement.”
# 
# Methodology: I used ChatGPT for suggestions on how to structure my pandas code more efficiently, to improve the correctness or clarity of formulas (like for groupby or probability estimation), and for advice on how to keep my simulations vectorized instead of relying on for loops. I also checked in for clarification on how to adhere to best practices when calculating aggregate statistics.
# 
# Debugging: I pasted in error messages and snippets of buggy code to ChatGPT to quickly identify and fix small issues, including selection/filtering bugs and mistakes in aggregations.
# 
# Writing, Review & Editing: I used ChatGPT to help make my written responses clearer and more concise, especially for questions where precise language and explanation were required. I sometimes asked for suggestions on rephrasing and organization to make sure my answers matched the rubric and were understandable.


<!-- END QUESTION -->

